wrappers/services.go

```diff
@@ -994,42 +1167,47 @@ func RestartServices(apps []*snap.AppInfo, explicitServices []string,
 	sysd := systemd.New(systemd.SystemMode, inter)
 
 	// Get service statuses for each of the apps
-	sts, err := internal.QueryServiceStatusMany(sysd, apps)
+	sysSvcs, usrSvcsMap, err := internal.QueryServiceStatusMany(apps, sysd)
 	if err != nil {
 		return err
 	}
 
-	for _, st := range sts {
-		unitName := st.ServiceUnitStatus().Name
-		unitActive := st.ServiceUnitStatus().Active
-		unitEnabled := st.IsEnabled()
+	// Handle restart of system services if scope was set
+	if flags.Scope != ServiceScopeUser {
+		if err := restartServicesByStatus(sysSvcs, explicitServices, flags, sysd, nil, tm); err != nil {
+			return err
+		}
+	}
 
-		// If the unit was explicitly mentioned in the command line, restart it
-		// even if it is disabled; otherwise, we only restart units which are
-		// currently enabled or running. Reference:
-		// https://forum.snapcraft.io/t/command-line-interface-to-manipulate-services/262/47
-		if !unitActive && !strutil.ListContains(explicitServices, unitName) {
-			if !flags.AlsoEnabledNonActive {
-				logger.Noticef("not restarting inactive unit %s", unitName)
-				continue
-			} else if !unitEnabled {
-				logger.Noticef("not restarting disabled and inactive unit %s", unitName)
+	// Handle restart of the user services if scope was set
+	if flags.Scope != ServiceScopeSystem {
+		// Get a list of the uids that we are affecting
+		uids, err := usersToUids(flags.Users)
+		if err != nil {
+			return err
+		}
+
+		for uid, stss := range usrSvcsMap {
+			// If specific users were specified, i.e self, then make sure we only
+			// restart services for that user
+			if len(uids) > 0 && uids[uid] == "" {
 				continue
 			}
-		}
 
-		var err error
-		timings.Run(tm, "restart-service", fmt.Sprintf("restart service %s", unitName), func(nested timings.Measurer) {
-			if flags.Reload {
-				err = sysd.ReloadOrRestart([]string{unitName})
-			} else {
-				// note: stop followed by start, not just 'restart'
-				err = sysd.Restart([]string{unitName})
+			var users []string
+			if uids[uid] != "" {
+				users = append(users, uids[uid])
+			}
+
+			// Create a new client, only targeting that user
+			cli, err := newUserServiceClient(users, inter)
```

Comment:
wouldn't it make sense to have a variant of this take uids directly to avoid the user -> uid -> user back and forth?

Reply:
super nice suggestion, added